File: src/python/auxiliares.py
from typing import *



BIG_NUMBER = 1e10

def error(sol: List[Tuple], x, y) -> float:
   res = 0
   i = 0
   a = sol[len(sol)-2]
   b = sol[-1]
   m = (b[1]-a[1])/(b[0]-a[0])
   c = a[1]-(m*a[0]) 
   while i<len(x) and x[i]<=b[0]: 
      if x[i]>=a[0]:
         res += abs(y[i]-(m*x[i]+c))
      i+=1
   return res

File: src/cpp/auxiliares.cpp
#include "pwl.h"
#include <iostream>
#include <algorithm>
using namespace std;


vector<float> grid(vector<float> x, int m){
    // precondicion: x_i < x_j para todo i < j
    vector<float> res ={};
    float min = *min_element(x.begin(), x.end());
    float max = *max_element(x.begin(), x.end());
    float k = (max-min)/(m-1);
    
    for (int i = 0; i < m; i+=1){
        res.push_back(min+k*i);
            
    }
    return res;  
};

float error(vector<tuple<float, float>> sol, vector<float> x, vector<float> y){
    
    float res = 0;
    int i = 0;
    tuple<float, float> a = sol[sol.size()-2];
    tuple<float, float> b = sol[sol.size()-1];
   
    float m = ((get<1>(b)-get<1>(a))/(get<0>(b)-get<0>(a)));
    float c = get<1>(a) - m*get<0>(a);
    
    while (i<x.size() && x[i]<=get<0>(b)){
        
        if(x[i] >= get<0>(a)){
            res += abs(y[i] - (m*x[i] + c));
        }
        i+=1;
    }
    
    return res;
};